diff -up Bumblebee/src/module.c.modprobefix Bumblebee/src/module.c
--- Bumblebee/src/module.c.modprobefix	2016-06-22 13:12:10.444633176 -0400
+++ Bumblebee/src/module.c	2016-06-22 13:15:04.374158011 -0400
@@ -92,16 +92,18 @@ int module_load(char *module_name, char
  * @return 1 if the driver is succesfully unloaded, 0 otherwise
  */
 int module_unload(char *driver) {
-  if (module_is_loaded(driver) == 1) {
+  if (module_is_loaded("nvidia_drm") == 1) {
     int retries = 30;
-    bb_log(LOG_INFO, "Unloading %s driver\n", driver);
+    bb_log(LOG_INFO, "Unloading nvidia_drm driver\n");
     char *mod_argv[] = {
-      "rmmod",
-      driver,
+      "modprobe",
+      "-r",
+      "nvidia-drm",
+      "nvidia-modeset",
       NULL
     };
     bb_run_fork_wait(mod_argv, 10);
-    while (retries-- > 0 && module_is_loaded(driver) == 1) {
+    while (retries-- > 0 && module_is_loaded("nvidia_drm") == 1) {
       usleep(100000);
     }
     if (module_is_loaded(driver) == 1) {
@@ -109,6 +111,44 @@ int module_unload(char *driver) {
       return 0;
     }
   }
+
+  else if (module_is_loaded("nvidia_modeset") == 1) {
+    int retries = 30;
+    bb_log(LOG_INFO, "Unloading nvidia_modeset driver\n");
+    char *mod_argv[] = {
+      "modprobe",
+      "-r",
+      "nvidia-modeset",
+      NULL
+    };
+    bb_run_fork_wait(mod_argv, 10);
+    while (retries-- > 0 && module_is_loaded("nvidia_modeset") == 1) {
+      usleep(100000);
+    }
+    if (module_is_loaded(driver) == 1) {
+      bb_log(LOG_ERR, "Unloading %s driver timed out.\n", driver);
+      return 0;
+    }
+  }
+
+  else if (module_is_loaded(driver) == 1) {
+      int retries = 30;
+      bb_log(LOG_INFO, "Unloading %s driver\n", driver);
+      char *mod_argv[] = {
+        "modprobe",
+        "-r",
+        driver,
+        NULL
+      };
+      bb_run_fork_wait(mod_argv, 10);
+      while (retries-- > 0 && module_is_loaded(driver) == 1) {
+        usleep(100000);
+      }
+      if (module_is_loaded(driver) == 1) {
+        bb_log(LOG_ERR, "Unloading %s driver timed out.\n", driver);
+        return 0;
+      }
+  }
   return 1;
 }
 
