diff -uNr net.orig/mac80211/tx.c net/mac80211/tx.c
--- net.orig/mac80211/tx.c	2018-05-02 15:58:43.000000000 +0800
+++ net/mac80211/tx.c	2019-09-13 07:04:45.727359923 +0800
@@ -765,8 +765,12 @@
 	 * number, if we have no matching interface then we
 	 * neither assign one ourselves nor ask the driver to.
 	 */
-	if (unlikely(info->control.vif->type == NL80211_IFTYPE_MONITOR))
-		return TX_CONTINUE;
+	if (unlikely((info->flags & IEEE80211_TX_CTL_INJECTED) &&
+	   !(tx->sdata->u.mntr_flags & MONITOR_FLAG_COOK_FRAMES))) {
+		if (!ieee80211_has_morefrags(hdr->frame_control))
+			info->flags |= IEEE80211_TX_CTL_NO_ACK;
+ 		return TX_CONTINUE;
+	}
 
 	if (unlikely(ieee80211_is_ctl(hdr->frame_control)))
 		return TX_CONTINUE;
