diff -uNr primus.orig/primusrun primus/primusrun
--- primus.orig/primusrun	2019-09-27 09:21:05.000000000 +0800
+++ primus/primusrun	2019-09-27 09:22:47.213906362 +0800
@@ -26,11 +26,12 @@
 # export PRIMUS_libGLd=${PRIMUS_libGLd:-'/usr/$LIB/libGL.so.1'}
 
 # Directory containing primus libGL
-PRIMUS_libGL=${PRIMUS_libGL:-$(dirname `readlink -ne $0`)/'$LIB'}
+PRIMUS_libGL='/usr/$LIB/primus'
 
 # On some distributions, e.g. on Ubuntu, libnvidia-tls.so is not available
 # in default search paths.  Add its path manually after the primus library
 # PRIMUS_libGL=${PRIMUS_libGL}:/usr/lib/nvidia-current:/usr/lib32/nvidia-current
+export __GLVND_DISALLOW_PATCHING=1
 
 # Mesa drivers need a few symbols to be visible
 # export PRIMUS_LOAD_GLOBAL=${PRIMUS_LOAD_GLOBAL:-'libglapi.so.0'}
